Solid Earth Tides correction

vbrancat opened the pull request:
This PR adds the Solid Earth Tides (SET) to the list of Look UP Table (LUT) corrections given to the `geocodeSlc` cor module. SET are computed using `pySolid`. The generated LUT in radar coordinates along slant range and azimuth are also added to the CSLC product.

Note that `pySolid` is not added as a dependency as the next `isce3` conda release should already support this dependency.

scottstanie commented on src/compass/utils/geometry_utils.py:
```diff
@@ -0,0 +1,166 @@
+
+
+def get_unit_vector4component_of_interest(los_inc_angle, los_az_angle, comp='enu2los', horz_az_angle=None):
+    """Get the unit vector for the component of interest.
+    Parameters: los_inc_angle - np.ndarray or float, incidence angle from vertical, in the unit of degrees
+                los_az_angle  - np.ndarray or float, azimuth angle of the LOS vector from the ground to the SAR platform
+                                measured from the north with anti-clockwise direction as positive, in the unit of degrees
+                comp          - str, component of interest, choose among the following values:
+                                enu2los, en2los, hz2los, u2los, up2los, orb(it)_az, vert, horz
+                horz_az_angle - np.ndarray or float, azimuth angle of the horizontal direction of interest
+                                measured from the north with anti-clockwise direction as positive, in the unit of degrees
+    Returns:    unit_vec      - list(np.ndarray/float), unit vector of the ENU component for the component of interest
+    """
```
```suggestion
    """Get the unit vector for the component of interest.

    Parameters
    ----------
    los_inc_angle : np.ndarray or float
        incidence angle from vertical, in the unit of degrees
    los_az_angle : np.ndarray or float
        azimuth angle (in degrees) of the LOS vector from the ground to the SAR platform
        measured from the north with anti-clockwise direction as positive
    comp : str
        component of interest, choose among the following values:
        enu2los, en2los, hz2los, u2los, up2los, orb(it)_az, vert, horz
    horz_az_angle : np.ndarray or float
        azimuth angle (in degrees) of the horizontal direction of interest
        measured from the north with anti-clockwise direction as positive

    Returns
    -------
    unit_vec : list[np.ndarray/float]
        unit vector of the ENU component for the component of interest
    """
```

scottstanie commented on src/compass/utils/geometry_utils.py:
```diff
@@ -0,0 +1,166 @@
+
+
+def calc_azimuth_from_east_north_obs(east, north):
+    """Calculate the azimuth angle of the given horizontal observation (in east and north)
+    Parameters: east     - float,  eastward motion
+                north    - float, northward motion
+    Returns:    az_angle - float, azimuth angle in degree
+                           measured from the north with anti-clockwise as positive
+    """
```
```suggestion
    """Calculate the azimuth angle of the given horizontal observation (in east and north)

    Parameters
    ----------
    east : float
        eastward motion
    north : float
        northward motion

    Returns
    -------
    az_angle : float
        azimuth angle (in degrees) measured from the north
        with anti-clockwise as positive
    """
```

scottstanie commented on src/compass/utils/geometry_utils.py:
```diff
@@ -0,0 +1,166 @@
+
+
+def en2az(v_e, v_n, orb_az_angle):
+    """Project east/north motion into the radar azimuth direction.
+    Parameters: v_e          - np.ndarray or float, displacement in east-west   direction, east  as positive
+                v_n          - np.ndarray or float, displacement in north-south direction, north as positive
+                orb_az_angle - np.ndarray or float, azimuth angle of the SAR platform along track/orbit direction
+                               measured from the north with anti-clockwise direction as positive, in the unit of degrees
+                               orb_az_angle = los_az_angle + 90 for right-looking radar.
+    Returns:    v_az         - np.ndarray or float, displacement in azimuth direction,
+                               motion along flight direction as positive
+    """
```
```suggestion
    """Project east/north motion into the radar azimuth direction.

    Parameters
    ----------
    v_e : np.ndarray or float
        displacement in east-west direction, east as positive
    v_n : np.ndarray or float
        displacement in north-south direction, north as positive
    orb_az_angle : np.ndarray or float
        azimuth angle (in degrees) of the SAR platform along track/orbit direction
        measured from the north with anti-clockwise direction as positive
        orb_az_angle = los_az_angle + 90 for right-looking radar.

    Returns
    -------
    v_az : np.ndarray or float
        displacement in azimuth direction.
        motion along flight direction as positive
    """
```

scottstanie commented on src/compass/utils/lut.py:
```diff
@@ -117,4 +132,202 @@ def compute_geocoding_correction_luts(burst, dem_path,
+                     heading_angle_raster=heading_raster)
+
+
+def open_raster(filename, band=1):
```
Seems like this might be useful in `utils/helpers.py`, since a quick search shows it might be useful in 5 other places:

```
src/compass/utils/elevation_antenna_pattern.py:    slc_in = gdal.Open(path_slc_vrt, gdal.GA_ReadOnly)
src/compass/utils/helpers.py:        gdal.Open(dem_path, gdal.GA_ReadOnly)
src/compass/utils/lut.py:    ds = gdal.Open(filename, gdal.GA_ReadOnly)
src/compass/utils/range_split_spectrum.py:    in_ds = gdal.Open(burst_path, gdal.GA_ReadOnly)
src/compass/utils/raster_polygon.py:        ds = gdal.Open(dataset_path, gdal.GA_ReadOnly)
src/compass/utils/validate_cslc.py:    dataset = gdal.Open(h5_gdal_path, gdal.GA_ReadOnly)
```

vbrancat replied:
I have moved the function in `helpers.py`. However, for some of the functions you have listed , we just read the geotranformation. So I have not recycled `open_raster` for that cases.

scottstanie reviewed (commented):
Still testing things out, but just submitting these suggestions to try and make it easy for you to convert the `geometry_utils.py` docstrings to numpydoc style so it matches the rest of the repo

vbrancat commented on src/compass/utils/geometry_utils.py:
```diff
@@ -0,0 +1,166 @@
+    return head_angle
+
+
+def heading2azimuth_angle(head_angle, look_direction='right'):
```
It would be very handy when working with ISCE2. The definition of this angle seems to be different between ISCE2 and ISCE3

seongsujeong reviewed (commented):
Thanks @vbrancat for the PR. I've left very minor comments here.
Let's discuss about how to convert meters -> seconds in azimuth. - Seongsu

vbrancat commented on src/compass/utils/lut.py:
```diff
@@ -3,6 +3,14 @@
+from compass.utils.helpers import open_raster
+from scipy.interpolate import RegularGridInterpolator as RGI
+from osgeo import gdal
+from skimage.transform import resize
```
I believe those dependencies will be added on the isce3 side. Also there we are trying to add the SET more or less in the same way we are doing here. That is why I have not added them in `requirements.txt`

LiangJYu replied:
`docker/specifile.txt` will need to be updated when we get an updated public isce3 release with the new packages.

scottstanie reviewed (commented):
With version of scipy that are older than 1.10 (maybe 1.9 changes this?), you get an error during SET interpolation: `ValueError: The points in dimension 0 must be strictly ascending`

```python-console
In [22]: import scipy

In [23]: scipy.__version__
Out[23]: '1.8.1'
```

See below:
<details>

```python-console
In [24]: fname = "data/S1B_IW_SLC__1SDV_20191002T135146_20191002T135213_018297_022770_1CDE.SAFE"
    ...: from compass.utils import lut
    ...: import isce3
    ...: import s1reader
    ...:
    ...: fname = "data/S1B_IW_SLC__1SDV_20191002T135146_20191002T135213_018297_022770_1CDE.SAFE"
    ...: orbit_file = s1reader.get_orbit_file_from_dir(fname, "data/orbits/")
    ...: burst = s1reader.load_bursts(fname, orbit_file, 1)[0]
    ...:
    ...: dem_raster = isce3.io.Raster("dem.tif")
    ...: ellipsoid = isce3.core.make_projection(dem_raster.get_epsg()).ellipsoid
    ...: output_path = "test_set"
    ...: rg_step, az_step = 200, 0.25
    ...:
    ...: geom_files = lut.compute_rdr2geo_rasters(burst, ellipsoid, dem_raster, output_path,
    ...:                                         #  rg_step, az_step)
    ...:                                         rg_step * 10, az_step * 10)
    ...:
    ...: from compass.utils.helpers import open_raster
    ...: lat, lon, inc_angle, head_angle = [open_raster(f)[0] for f in geom_files]
    ...:
    ...: rg_set_temp, az_set_temp = lut.solid_earth_tides(burst, lat, lon, inc_angle, head_angle)
measurement directory NOT found in data/S1B_IW_SLC__1SDV_20191002T135146_20191002T135213_018297_022770_1CDE.SAFE, continue with metadata only.
journal:
 -- DEM EPSG: 4326
 -- Output EPSG: 4326
journal:
 -- Processing block: 0
 --   - line start: 0
 --   - line end  : 2
 --   - dopplers near mid far: 0 0 0
journal: West limit may be insufficient for global height range
journal: East limit may be insufficient for global height range
journal: South limit may be insufficient for global height range
journal: North limit may be insufficient for global height range
journal:
 -- Actual DEM bounds used:
 -- Top Left: -117.247 34.6403
 -- Bottom Right: -116.266 34.3283
 -- Spacing: 0.000277778 -0.000277778
 -- Dimensions: 3534 1124
journal: Total convergence: 50 out of 50
PYSOLID: ----------------------------------------
PYSOLID: datetime: 2019-10-02T13:51:47.886427
PYSOLID: SNWE: (34.80321797417274, 34.228217974172736, -117.3471080758785, -115.0471080758785)
SOLID  : calculate solid Earth tides in east/north/up direction
SOLID  : shape: (25, 100), step size: 0.0230 by 0.0230 deg
SOLID  : calculating / writing data to txt file: /Users/staniewi/repos/COMPASS/solid.txt
PYSOLID: read data from text file: /Users/staniewi/repos/COMPASS/solid.txt
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <cell line: 22>                                                                               │
│                                                                                                  │
│ /Users/staniewi/repos/COMPASS/src/compass/utils/lut.py:237 in solid_earth_tides                  │
│                                                                                                  │
│   234 │   pts_src = (np.flipud(lat_geo_array), lon_geo_array)                                    │
│   235 │   pts_dst = (lat_radar_grid.flatten(), lon_radar_grid.flatten())                         │
│   236 │                                                                                          │
│ ❱ 237 │   rdr_set_e = resample_set(set_e, pts_src, pts_dst).reshape(                             │
│   238 │   │   lat_radar_grid.shape)                                                              │
│   239 │   rdr_set_n = resample_set(set_n, pts_src, pts_dst).reshape(                             │
│   240 │   │   lat_radar_grid.shape)                                                              │
│                                                                                                  │
│ /Users/staniewi/repos/COMPASS/src/compass/utils/lut.py:347 in resample_set                       │
│                                                                                                  │
│   344 │                                                                                          │
│   345 │   # Flip tide displacement component to be consistent with flipped latitudes             │
│   346 │   geo_tide = np.flipud(geo_tide)                                                         │
│ ❱ 347 │   rgi_func = RGI(pts_src, geo_tide, method='nearest',                                    │
│   348 │   │   │   │      bounds_error=False, fill_value=0)                                       │
│   349 │   rdr_tide = rgi_func(pts_dest)                                                          │
│   350 │   return rdr_tide                                                                        │
│                                                                                                  │
│ /Users/staniewi/miniconda3/envs/mapping/lib/python3.10/site-packages/scipy/interpolate/_interpol │
│ ate.py:2485 in __init__                                                                          │
│                                                                                                  │
│   2482 │   │                                                                                     │
│   2483 │   │   for i, p in enumerate(points):                                                    │
│   2484 │   │   │   if not np.all(np.diff(p) > 0.):                                               │
│ ❱ 2485 │   │   │   │   raise ValueError("The points in dimension %d must be strictly "           │
│   2486 │   │   │   │   │   │   │   │    "ascending" % i)                                         │
│   2487 │   │   │   if not np.asarray(p).ndim == 1:                                               │
│   2488 │   │   │   │   raise ValueError("The points in dimension %d must be "                    │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: The points in dimension 0 must be strictly ascending

```

</details>

~But also after I upgrade to scipy 1.10 and it stops erroring, it looks like something's getting messed up: all the SET corrections are 0 right now~ Edit: I was testing it incorrectly with one of my suggested changes, the range seems fine as is.

Theres a `# TO DO: add azimuth SET to LUT`. Is that because we're not sure if the azimuth output is correct right now? 
I see it's giving values of about -0.02013801 for my example data for the `az_set`.  that in seconds? Does that seem right if the largest ENU SET value is about 4 cm? 
```
ipdb> p np.max([np.abs(set_e).max(), np.abs(set_n).max(), np.abs(set_u).max()])
0.0471502506
```

scottstanie commented:
A separate question I thoguht of: Is there any plan to separately save the corrections we apply (so a user could un-apply one of `geometrical_steer_doppler, bistatic_delay, az_fm_mismatch, tides`), or are we just saving the cumulative one?

LiangJYu commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,230 @@ def compute_geocoding_correction_luts(burst, dem_path,
                                                           az_step=az_step)
 
-    return geometrical_steering_doppler, bistatic_delay, az_fm_mismatch
+    # Get solid Earth tides on a very coarse grid
+    # Run rdr2geo on a very coarse grid
+    compute_rdr2geo_rasters(burst, ellipsoid, dem_raster, output_path,
+                            rg_step * 10, az_step * 10)
+
+    # Open rdr2geo layers and feed them to SET computation
+    lat, _ = open_raster(f'{output_path}/y.rdr')
+    lon, _ = open_raster(f'{output_path}/x.rdr')
+    inc_angle, _ = open_raster(f'{output_path}/incidence_angle.rdr')
+    head_angle, _ = open_raster(f'{output_path}/heading_angle.rdr')
+
+    # compute Solid Earth Tides (using pySolid)
+    rg_set_temp, az_set_temp = solid_earth_tides(burst, lat, lon,
+                                                 inc_angle, head_angle)
+
+    # Resize SET to the size of the correction grid
+    out_shape = bistatic_delay.data.shape
+    kwargs = dict(order=1, mode='edge', anti_aliasing=True,
+                  preserve_range=True)
+    rg_set = resize(rg_set_temp, out_shape, **kwargs)
+    az_set = resize(az_set_temp, out_shape, **kwargs)
```
In line with what @scottstanie's [comment](https://github.com/opera-adt/COMPASS/pull/91/files#r1111324064), unless there's utility elsewhere, the lines preceding and following the call the `solid_earth_tides()` could also be in `solid_earth_tides()`.

vbrancat replied:
Uhm, we will be using the rdr2geo layers and the corresponding function that computes them for the computation of the troposphere delay using weather model. Probably, we can wait for that part of code to come out and then get back to this comment?

vbrancat commented on src/compass/utils/helpers.py:
```diff
@@ -318,3 +318,27 @@ def burst_bboxes_from_db(burst_ids, burst_db_file=None, burst_db_conn=None):
+    ds = gdal.Open(filename, gdal.GA_ReadOnly)
+    raster = ds.GetRasterBand(band).ReadAsArray()
+    data_type = ds.GetRasterBand(band).DataType
+    return raster, data_type
```
You are right. There was no use to return the data type so I removed it from the returning arguments of the function

vbrancat commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,230 @@ def compute_geocoding_correction_luts(burst, dem_path,
+    '''
+
+    # Get radar and doppler grid
+    width_rdr_grid, length_rdr_grid = [vec.size for vec in
+                                       burst._steps_to_vecs(rg_step, az_step)]
+
+    rdr_grid = isce3.product.RadarGridParameters(
+        burst.as_isce3_radargrid().sensing_start,
+        burst.wavelength,
+        1.0 / az_step,
+        burst.starting_range,
+        rg_step,
+        isce3.core.LookSide.Right,
+        length_rdr_grid,
+        width_rdr_grid,
+        burst.as_isce3_radargrid().ref_epoch
+    )
```
I think it will come handy to have a method like that in the `s1-reader`. Will wait for the PR to be merged to refactor this part of the code

scottstanie commented on src/compass/utils/lut.py:
```diff
@@ -3,6 +3,14 @@
 '''
 import os
 import isce3
+import numpy as np
+import pysolid
+
+from compass.utils.geometry_utils import enu2los, en2az
+from compass.utils.helpers import open_raster
+from scipy.interpolate import RegularGridInterpolator as RGI
+from osgeo import gdal
+from skimage.transform import resize
```
Just curious, how did you get that ordering @LiangJYu ? When I run `isort` on it in VSCode, it produces this

```python
import os

import isce3
import numpy as np
import pysolid
from osgeo import gdal
from scipy.interpolate import RegularGridInterpolator as RGI
from skimage.transform import resize

from compass.utils.geometry_utils import en2az, enu2los
from compass.utils.helpers import open_raster
```

LiangJYu replied:
In my head, I order without regard `import` and `from` and sort only by library name. I like it avoids partitioning imports from the same library into different blocks. There's no official guidance in PEP8. I think my implementation is consistent the [google style guide](https://google.github.io/styleguide/pyguide.html#313-imports-formatting). Clearly `isort` disagrees. 😄

vbrancat commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,230 @@ def compute_geocoding_correction_luts(burst, dem_path,
+                                         threshold=1.0e8)
+
+    # Get the rdr2geo raster needed for SET computation
+    topo_output = {'x': (True, gdal.GDT_Float64),
+                   'y': (True, gdal.GDT_Float64),
+                   'incidence_angle': (True, gdal.GDT_Float32),
+                   'heading_angle': (True, gdal.GDT_Float32)}
```
No, I believe we have no plan to tweak these Booleans. Are you suggesting to strip out the "True" from the dictionary?

scottstanie replied:
yep exactly, then if could just be `for fname, dtype in topo_output.items()` in the for loop without checking for `enabled`

vbrancat commented:
> A separate question I thoguht of: Is there any plan to separately save the corrections we apply (so a user could un-apply one of `geometrical_steer_doppler, bistatic_delay, az_fm_mismatch, tides`), or are we just saving the cumulative one?

The individual corrections are saved separately inside the product in the `corrections` group.

vbrancat commented:
I think I have addressed all your comments. I would appreciate if you all can have another look at the PR. In the meantime, @seongsujeong is working on addressing the conversion of the azimuth component of SET from meters to seconds.

scottstanie reviewed (approved):
LGTM!

hfattahi reviewed (commented):
This PR looks good overall. It will be great to see the impact on real data to make sure about the sign of the correction. Do we have any results? Or any evidence that the calculations are correct?

vbrancat commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,246 @@ def compute_geocoding_correction_luts(burst, dem_path,
+    # Run pySolid and get SET in ENU coordinate system
+    (set_e,
+     set_n,
+     set_u) = pysolid.calc_solid_earth_tides_grid(burst.sensing_start, atr,
+                                                  display=False, verbose=True)
```
Yeah, I have noticed. I ran pySolid on a very coarse grid so this should not affect the runtime dramatically

LiangJYu commented on src/compass/utils/helpers.py:
```diff
@@ -318,3 +318,25 @@ def burst_bboxes_from_db(burst_ids, burst_db_file=None, burst_db_conn=None):
+
+def open_raster(filename, band=1):
+    '''
+    Return band as numpy array from gdal-friendly raster
```
```suggestion
    Return band as numpy array from gdal-friendly raster
    
```
formatting nit

LiangJYu commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,246 @@ def compute_geocoding_correction_luts(burst, dem_path,
+        in meters. These corrections need to be added to the slC tagged azimuth
+        and slant range times.
     '''
+    # Get DEM raster
```
```suggestion
    if not os.path.isfile(dem_path):
        raise FileNotFoundError(f'Cannot find the dem file: {dem_path}')
    # Get DEM raster
```
Check `dem_path` before raster init.

LiangJYu commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,246 @@ def compute_geocoding_correction_luts(burst, dem_path,
 
-    if not os.path.exists(dem_path):
+    # Compute azimuth FM-rate mismatch
+    if not os.path.isfile(dem_path):
         raise FileNotFoundError(f'Cannot find the dem file: {dem_path}')
```
```suggestion
```
Cove check above before DEM raster init

LiangJYu reviewed (approved):
LGTM aside from a couple of small comments.

seongsujeong commented on src/compass/utils/lut.py:
```diff
@@ -103,18 +114,230 @@ def compute_geocoding_correction_luts(burst, dem_path,
+        'Y_FIRST': lat_start,
+        'X_STEP': 0.023,
+        'Y_STEP': 0.023
+    }
```
Tested the burst `t001_000689_iw1`, whose approximate lat / lon bounding box is as shown in the PYSOLID log below. The solid tide in ENU is as in the screenshot below. Looks like SET correction works well on antemeridian case.

```
PYSOLID: ----------------------------------------
PYSOLID: datetime: 2022-02-25T18:32:18.459594
PYSOLID: SNWE: (64.65123432456033, 64.07623432456033, 179.9, 182.20000000000002)
SOLID  : calculate solid Earth tides in east/north/up direction
SOLID  : shape: (25, 100), step size: 0.0230 by 0.0230 deg
SOLID  : calculating / writing data to txt file: /Users/jeong/Documents/OPERA_SCRATCH/CSLC/SET_antimeridian/solid.txt
PYSOLID: read data from text file: /Users/jeong/Documents/OPERA_SCRATCH/CSLC/SET_antimeridian/solid.txt
```

<img width="752" alt="Screenshot 2023-03-02 at 12 10 41" src="https://user-images.githubusercontent.com/27862199/222541276-8293f6fe-380b-4e41-8818-4ffca4031db7.png">

scottstanie commented on src/compass/utils/helpers.py:
```diff
@@ -340,3 +341,55 @@ def open_raster(filename, band=1):
+                  f'the number of descriptions ' \
+                  f'{len(data_list)} != {len(descriptions)}'
+        error_channel.log(err_str)
+        raise ValueError(err_str)
```
Not sure if we want to change these since we're in the process of converting all logging to python, but Michael pointed out that the `raise ...` statements in COMPASS that occur after an `error_channel.log` call will never be reached. The error channel is "fatal" aka it raises it's own error first:

```python
In [2]: import journal
In [3]: c = journal.error('test1')

In [5]: c.log("bad")
journal: bad
---------------------------------------------------------------------------
ApplicationError                          Traceback (most recent call last)
Cell In[5], line 1
----> 1 c.log("bad")

ApplicationError: test1: application error
```

seongsujeong commented on docker/Dockerfile:
```diff
@@ -37,9 +37,9 @@ SHELL ["conda", "run", "-n", "COMPASS", "/bin/bash", "-c"]
+RUN curl -sSL https://github.com/opera-adt/s1-reader/archive/main.tar.gz -o s1_reader_src.tar.gz &&\
     tar -xvf s1_reader_src.tar.gz &&\
-    ln -s s1-reader-0.1.5 s1-reader &&\
+    ln -s s1-reader-main s1-reader &&\
```
This is better for CI to take the changes in `s1-reader` in to account, but I think we'll need to hard code its version in the upcoming deliveries. @vbrancat @LiangJYu Any opinions?

seongsujeong reviewed (approved):
The main part of the SET correction looks good to me. I see some version issue for `s1-reader` in Dockerfile, but I understand this is necessary at this point to use the recent changes in the reader.

Let's revisit the Docker thing when we start to work on the delivery.
